src/communication/communicator.hpp

```diff
@@ -45,31 +48,71 @@ class communicator {
 
     communicator() {}
 
-    explicit communicator(gid_partition_type cell_gid_partition):
-        cell_gid_partition_(cell_gid_partition)
-    {}
-
-    cell_local_size_type num_groups_local() const
-    {
-        return cell_gid_partition_.size();
-    }
-
-    void add_connection(connection con) {
-        EXPECTS(is_local_cell(con.destination().gid));
-        connections_.push_back(con);
-    }
-
-    /// returns true if the cell with gid is on the domain of the caller
-    bool is_local_cell(cell_gid_type gid) const {
-        return algorithms::in_interval(gid, cell_gid_partition_.bounds());
-    }
+    explicit communicator(const recipe& rec, const domain_decomposition& dom_dec) {
```

Comment:
I believe the goal is still to have the description of the domain decomposition be a separate type from the object that actually performs the decomposition (certainly this is what I was advocating before, and I _think_ I've convinced @bcumming).

The `communicator` object will still need to get the source gid list per target and will need the gid per group info to make the event queues, so I think we're thereby constrained to an interface which is at least equivalent to _recipe_ × _domain decomposition_ → _communicator_